src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1645 @@
+        ``(c, factors)`` where `c` is the content of the coefficients and
+        factors is a list of ``(poly, exp)`` pairs.
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> x = R.gen()
+            >>> f = (x + 1) * (x + 2)
+            >>> f.factor_squarefree()
+            (fmpz_mod(1, 163), [(x^2 + 3*x + 2, 1)])
+            >>> f = (x + 1) * (x + 2)**5
+            >>> f.factor_squarefree()
+            (fmpz_mod(1, 163), [(x + 1, 1), (x + 2, 5)])
+        """
+        cdef fmpz_mod_poly_factor_t fac
+        cdef int i
+
+        if not self.ctx.is_prime():
+            raise NotImplementedError("factor_squarefree algorithm assumes that the modulus is prime")
+
+        fmpz_mod_poly_factor_init(fac, self.ctx.mod.val)
+        fmpz_mod_poly_factor_squarefree(fac, self.val, self.ctx.mod.val)
+
+        res = [0] * fac.num
+
+        cdef fmpz_mod_poly u
+        for i in range(fac.num):
+            u = self.ctx.new_ctype_poly()
+            fmpz_mod_poly_set(u.val, &fac.poly[i], self.ctx.mod.val)
+            exp = fac.exp[i]
+            res[i] = (u, exp)
+        return self.leading_coefficient(), res
+
+    def factor(self, algorithm=None):
+        """
+        Factors self into irreducible factors, returning a tuple
+        ``(c, factors)`` where `c` is the content of the coefficients and
+        factors is a list of ``(poly, exp)`` pairs.
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> x = R.gen()
+            >>> f = 6*x**4 + 7*x**3 + 7*x**2 + 8*x + 6
+            >>> f.factor()
+            (fmpz_mod(6, 163), [(x^4 + 137*x^3 + 137*x^2 + 110*x + 1, 1)])
+            >>> f = (x + 1)**3 * (x + 2)
+            >>> f.factor()
+            (fmpz_mod(1, 163), [(x + 1, 3), (x + 2, 1)])
+        """
+        cdef fmpz_mod_poly_factor_t fac
+        cdef int i
+
+        if not self.ctx.is_prime():
+            raise NotImplementedError("factor algorithm assumes that the modulus is prime")
+                 
+        fmpz_mod_poly_factor_init(fac, self.ctx.mod.val)
+        if algorithm == None:
+            fmpz_mod_poly_factor(fac, self.val, self.ctx.mod.val)
+        elif algorithm == "cantor_zassenhaus":
+            fmpz_mod_poly_factor_cantor_zassenhaus(fac, self.val, self.ctx.mod.val)
+        elif algorithm == "kaltofen_shoup":
+            fmpz_mod_poly_factor_kaltofen_shoup(fac, self.val, self.ctx.mod.val)
+        elif algorithm == "berlekamp":
+            fmpz_mod_poly_factor_berlekamp(fac, self.val, self.ctx.mod.val)
+        else:
+            raise ValueError("unknown algorithm")
+        
+        res = [0] * fac.num
+
+        cdef fmpz_mod_poly u
+        for i in range(fac.num):
+            u = self.ctx.new_ctype_poly()
+            fmpz_mod_poly_set(u.val, &fac.poly[i], self.ctx.mod.val)
+            exp = fac.exp[i]
+            res[i] = (u, exp)
+        return self.leading_coefficient(), res
+
+    def roots(self, multiplicities=True):
```

Comment:
I generally prefer separate functions rather than a boolean flag

Reply:
So I think the default should be with multiplicities, so you think a `root_without_multiplicities()` function would be better?

Reply:
Coming up with good names is hard :)

Maybe `root_set`?

In this situation I always ant to check what other systems do...

Reply:
Root set is definitely better than my suggestion. Haha. For SageMath, they have a bool flag.

Reply:
> For SageMath, they have a bool flag.

Yeah, so does SymPy:
```
In [1]: roots((x**2 - 1)**2)
Out[1]: {-1: 2, 1: 2}

In [2]: roots((x**2 - 1)**2, multiple=True)
Out[2]: [-1, -1, 1, 1]
```
I think it is better though if a function has a clear return type and so if you want to return different types then you should have different function names. Also it is clearer just to write `roots_multiple(...)` rather than e.g. `roots(..., multiple=True)`.

Reply:
So if I thought not multiplicities was default then I'd agree but I think it's fine for a user to essentially never know this is an option as worst case they filter out the multiplicities themselves in python? The bool flag here is really only for people who want it but it's not needed. 

Agree about consistent return types tho. Not sure what's best

Reply:
I'm not sure what's best either really.

It is usually better just to have separate functions if there is a need to provide different options. It can also be good just to refuse to provide any options that are potentially redundant.

For python-flint right now I would aim to avoid (postpone?) providing anything that is not strictly necessary because it is a lot of work just to provide the strictly necessary things when you consider that all types (`fmpz_poly` etc) should ideally support the same usage.

Reply:
Maybe we could just remove the option to not compute multiplicities then? I'm not at my laptop now but I can edit it out tomorrow.

Reply:
We can leave it for now.